File: solutions/ruby/yacht/1/yacht.rb
class Yacht
  attr_reader :score

  def initialize(scores, category)
    @score = case category
    when 'ones' then scores.select{|score| score == 1}.sum
    when 'twos' then scores.select{|score| score == 2}.sum
    when 'threes' then scores.select{|score| score == 3}.sum
    when 'fours' then scores.select{|score| score == 4}.sum
    when 'fives' then scores.select{|score| score == 5}.sum
    when 'sixes' then scores.select{|score| score == 6}.sum
    when 'full house' then scores.tally.values.sort == [2, 3] ? scores.sum : 0
    when 'four of a kind' then scores.tally.values.sort.last >= 4 ? scores.tally.select{|k,v| v >= 4}.keys[0] * 4 : 0
    when 'little straight' then scores.sort == [1,2,3,4,5] ? 30 : 0
    when 'big straight' then scores.sort == [2,3,4,5,6] ? 30 : 0
    when 'choice' then scores.sum
    when 'yacht' then scores.all? {|score| score == scores[0]} ? 50 : 0
    end
  end
end

File: solutions/python/yacht/1/yacht.py
import collections
import enum


class YatchEnum(enum.Enum):
    YACHT = enum.auto()
    ONES = enum.auto()
    TWOS = enum.auto()
    THREES = enum.auto()
    FOURS = enum.auto()
    FIVES = enum.auto()
    SIXES = enum.auto()
    FULL_HOUSE = enum.auto()
    FOUR_OF_A_KIND = enum.auto()
    LITTLE_STRAIGHT = enum.auto()
    BIG_STRAIGHT = enum.auto()
    CHOICE = enum.auto()


# Only for testing purposes
YACHT = YatchEnum.YACHT
ONES = YatchEnum.ONES
TWOS = YatchEnum.TWOS
THREES = YatchEnum.THREES
FOURS = YatchEnum.FOURS
FIVES = YatchEnum.FIVES
SIXES = YatchEnum.SIXES
FULL_HOUSE = YatchEnum.FULL_HOUSE
FOUR_OF_A_KIND = YatchEnum.FOUR_OF_A_KIND
LITTLE_STRAIGHT = YatchEnum.LITTLE_STRAIGHT
BIG_STRAIGHT = YatchEnum.BIG_STRAIGHT
CHOICE = YatchEnum.CHOICE


def score(dice: list[int], category: YatchEnum) -> int:
    for index, _category in enumerate([
            YatchEnum.ONES,
            YatchEnum.TWOS,
            YatchEnum.THREES,
            YatchEnum.FOURS,
            YatchEnum.FIVES,
            YatchEnum.SIXES]):
        if _category == category:
            return sum(filter(lambda score: score == index + 1, dice))

    if category == YatchEnum.FULL_HOUSE:
        return sum(dice) if sorted(collections.Counter(dice).values()) == [2, 3] else 0

    if category == YatchEnum.CHOICE:
        return sum(dice)

    if category == YatchEnum.FOUR_OF_A_KIND:
        faces = set(key for key, value in collections.Counter(
            dice).items() if value >= 4)

        if len(faces) == 0:
            return 0

        face = faces.pop()

        return face * 4

    if category == YatchEnum.LITTLE_STRAIGHT:
        return 30 if sorted(dice) == [1, 2, 3, 4, 5] else 0

    if category == YatchEnum.BIG_STRAIGHT:
        return 30 if sorted(dice) == [2, 3, 4, 5, 6] else 0

    if category == YatchEnum.YACHT:
        return 50 if list(collections.Counter(dice).values()) == [5] else 0
